cartographer/mapping_3d/sparse_pose_graph.cc

```diff
@@ -423,6 +423,34 @@ void SparsePoseGraph::RunFinalOptimization() {
           .max_num_iterations());
 }
 
+void SparsePoseGraph::CalculateHistograms() {
+  translational_residual_ = common::Histogram();
+  for (const Constraint& constraint : constraints_) {
+    if (constraint.tag == Constraint::Tag::INTRA_SUBMAP) {
+      // in map frame
+      const cartographer::transform::Rigid3d& optimized_pose_map =
+          trajectory_nodes_.at(constraint.node_id).pose;
+      // in submap frame
+      const cartographer::transform::Rigid3d& original_pose_submap =
+          constraint.pose.zbar_ij;
+      // find the submap to map transform.
+      const cartographer::transform::Rigid3d& submap_transform =
+          optimized_submap_transforms_.at(constraint.submap_id.trajectory_id)
+              .at(constraint.submap_id.submap_index)
+              .pose;
+
+      // original pose in the map frame.
+      const cartographer::transform::Rigid3d original_pose_map =
```

Comment:
Might be more readable to combine the two optimized poses to an `optimized_node_to_submap`, then compare to the `node_to_submap_constraint`.